### PR TITLE
DEPS: Bump numpy to 1.13.3

### PR DESCRIPTION
**Pull request by h-vetinari**
#25227 is trying to add `pyproject.toml` which is necessary for the whole `pip`-machinery to work smoothly post v.19.0. However, this seems to be running into problems with a numpy version bump, and mood there was to bump numpy requirement to 1.13.x.

The bare minimum for the bump is implemented in that PR, but a version bump requires getting rid of a bunch of compat code (and CI adjustments), otherwise the cruft just accumulates (it took me three PRs to clean out until 1.12).

Scipy is goint to [bump to 1.13.3](https://github.com/pandas-dev/pandas/pull/25227#issuecomment-464128057) with the next release (due to cython issues, I believe), so I thought I'll directly take it to 1.13.3 here.

I tried redistributing the CI jobs as best as I could. Here's the before/after:

**EDIT: updated pins in https://github.com/pandas-dev/pandas/pull/25554#issuecomment-475881206**

CI job: line number | before | this PR
-------|-----------------------------|-------------------
`ci/deps/azure-27-compat.yaml:10:` | `numpy=1.12.0` | `numpy=1.13.3`
`ci/deps/azure-27-locale.yaml:10:` | `numpy=1.12.0` |  `numpy=1.14.*`
`ci/deps/azure-36-locale_slow.yaml:16:` | `numpy` | `numpy=1.15.*`
`ci/deps/azure-37-locale.yaml:15:` | `numpy` | `numpy`
`ci/deps/azure-37-numpydev.yaml:18:` | `numpy` | `numpy`
`ci/deps/azure-macos-35.yaml:14:` | `numpy=1.12.0` | `numpy=1.13.3`
`ci/deps/azure-windows-27.yaml:15:` | `numpy=1.12*` | `numpy=1.13.3`
`ci/deps/azure-windows-36.yaml:12:` | `numpy=1.14*` | `numpy=1.15.*`
`ci/deps/travis-27.yaml:20:` | `numpy=1.13*` | `numpy=1.14.*`
`ci/deps/travis-36-doc.yaml:23:` | `numpy=1.13*` | `numpy`
`ci/deps/travis-36-locale.yaml:15:` | `numpy` | `numpy`
`ci/deps/travis-36-slow.yaml:12:` | `numpy` | `numpy`
`ci/deps/travis-36.yaml:17:` | `numpy` | `numpy=1.15.*`
`ci/deps/travis-37.yaml:10:` | `numpy` | `numpy`

Matplotlib needs to be bumped as well (since current min 2.0.0 cannot be resolved by conda together with numpy 1.13.3), but it's only a tiny bump to ~2.0.2~ ~2.1.0~ ~2.1.1~ 2.2.2. In any case, the version spread should probably start to reflect that there's a bunch of mpl 3.0.x about, and so I also changed the spread a bit here:

CI job: line number | before | this PR
-------------|-------|-----------
`ci/deps/azure-27-locale.yaml:9:` | `matplotlib=2.0.0` | `matplotlib=2.2.2`
`ci/deps/azure-36-locale_slow.yaml:13:` | `matplotlib` | `matplotlib=3.0.*`
`ci/deps/azure-37-locale.yaml:12:` | `matplotlib` | `matplotlib`
`ci/deps/azure-macos-35.yaml:11:` | `matplotlib=2.2.0` | `matplotlib=2.2.3`
`ci/deps/azure-windows-27.yaml:13:` | `matplotlib=2.0.1` | `matplotlib=2.2.4`
`ci/deps/azure-windows-36.yaml:10:` | `matplotlib` | `matplotlib`
`ci/deps/travis-27.yaml:16:` | `matplotlib=2.2.2` | `matplotlib=2.2.*`
`ci/deps/travis-36-doc.yaml:17:` | `matplotlib` | `matplotlib`
`ci/deps/travis-36-locale.yaml:12:` | `matplotlib` | `matplotlib=3.0.1`
`ci/deps/travis-36-slow.yaml:10:` | `matplotlib` | `matplotlib`
`ci/deps/travis-36.yaml:14:` | `matplotlib` | `matplotlib=3.0.0`

It's probably wishful thinking that everything will pass straight away (especially when mpl versions change, see the code in `pandas/tests/plotting/test_datetimelike.py`), but oh well...

### CODE REVIEW

**Comment by rgommers** [comment]
> Scipy is goint to bump to 1.13.3 with the next release (due to cython issues, I believe), so I thought I'll directly take it to 1.13.3 here.

Not due to Cython issues, just to reduce the maintenance burden. See https://mail.python.org/pipermail/scipy-dev/2018-November/023181.html for rationale.

+1 for going to 1.13.3 as well. Matplotlib is doing a similar thing (actually they want to be a bit more agressive), and wants to standardize a common Python/NumPy version support policy for core packages (see https://github.com/scipy/scipy/issues/9610, we'll come back on that soon).

**Comment by jschendel** [comment]
`install.rst` also needs a small update:

https://github.com/pandas-dev/pandas/blob/615fbb3ce4bb171e8610c590a9f9899c60028d54/doc/source/install.rst#L227

File: ci/deps/travis-36.yaml
Comment (jreback) [inline ci/deps/travis-36.yaml]: i am -0 kn pinning this

**Review by jreback — CHANGES_REQUESTED** [review]

File: README.md
Comment (h-vetinari) [reply README.md]: This has been the oldest tested pytz for a while (also before #23062), but it's easy to forget updating `README.md`. It's the oldest in current testing:

```
(pandas-dev) C:\Users\<xxx>\pddev>findstr /L /S /N pytz ci/*
ci/deps\azure-27-compat.yaml:15:  - pytz=2013b
ci/deps\azure-27-locale.yaml:15:  - pytz
ci/deps\azure-27-locale.yaml:16:  - pytz=2013b
ci/deps\azure-36-locale_slow.yaml:21:  - pytz
ci/deps\azure-37-locale.yaml:20:  - pytz
ci/deps\azure-37-numpydev.yaml:6:  - pytz
ci/deps\azure-macos-35.yaml:19:  - pytz
ci/deps\azure-windows-27.yaml:19:  - pytz
ci/deps\azure-windows-36.yaml:19:  - pytz
ci/deps\travis-27.yaml:33:  - pytz=2013b
ci/deps\travis-36-doc.yaml:33:  - pytz
ci/deps\travis-36-locale.yaml:22:  - pytz
ci/deps\travis-36-slow.yaml:20:  - pytz
ci/deps\travis-36.yaml:25:  - pytz
ci/deps\travis-37.yaml:14:  - pytz
```

Also, 2013b is ancient (will also be bumped by #24942)

File: ci/deps/azure-36-locale_slow.yaml
Comment (h-vetinari) [reply ci/deps/azure-36-locale_slow.yaml]: will answer in top comment

File: ci/deps/travis-27.yaml
Comment (h-vetinari) [reply ci/deps/travis-27.yaml]: will answer in top comment

File: ci/deps/travis-36.yaml
Comment (h-vetinari) [reply ci/deps/travis-36.yaml]: will answer in top comment

**Review by h-vetinari — COMMENTED** [review]
@jreback 
Regarding pins, please have a look at the OP. Bumping numpy is always a good time to look at the distribution in the CI.

For numpy, there's currently 3x `1.12`, 2x `1.13`, 1x `1.14`, ZERO `1.15` and 7 unpinned.
I'm proposing to make this 3x `1.13`, 2x `1.14`, 2x `1.15` and 6 unpinned.

For matplotlib, there's currently `2.0.0`, `2.0.1`, `2.2.0`, `2.2.2` and 7 unpinned.
I'm proposing to make this `2.0.2` (new min), `2.1.0`, `2.2.0`, `2.2.*` (LTS!), `3.0.0`, `3.0.1`, `3.0.*`  and 4 unpinned.

**Comment by TomAugspurger** [comment]
I like @h-vetinari's proposed pins in 
https://github.com/pandas-dev/pandas/pull/25554#issue-258441536

**Comment by h-vetinari** [comment]
@jreback @TomAugspurger 
This should be done now. :)

**Review by TomAugspurger — COMMENTED** [review]
LGTM.

@h-vetinari if you're interested, I believe that this will let us delete the following on SparseArray / SparseSeries

- `__array_wrap__`
- `__array_prepare__`
- `__array_priority__`

Those are disabled when the class defines `__array_ufunc__`, which was started in NumPy 1.13

**Comment by h-vetinari** [comment]
@TomAugspurger 
That sounds like a separate PR?

@jreback 
PTAL

**Comment by h-vetinari** [comment]
> > @TomAugspurger: @h-vetinari if you're interested [...]

> @h-vetinari: @TomAugspurger That sounds like a separate PR?

I've got an initial draft prepared, but will wait until this one is merged.

**Comment by jreback** [comment]
this needs to wait until 0.24.2 is done and needs review

File: ci/deps/travis-36.yaml
Comment (jreback) [inline ci/deps/travis-36.yaml]: is there a reason this is pinned exactly to 3.0.0, which was superseded immedately by 3.0.1?

File: pandas/tests/plotting/common.py
Comment (jreback) [inline pandas/tests/plotting/common.py]: I would just bump mpl to 2.1 min to avoid all of these extra things

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: add the issue number (this one).

File: ci/deps/travis-36.yaml
Comment (h-vetinari) [reply ci/deps/travis-36.yaml]: There were almost three months between 3.0.0 and 3.0.1, see https://github.com/matplotlib/matplotlib/releases.

I just tried to spread some pins throughout the 3.x Series, see OP.

File: pandas/tests/plotting/common.py
Comment (h-vetinari) [reply pandas/tests/plotting/common.py]: ok, will do, but the special casing for 2.2.0 will still be necessary, unless we bump to 2.2.2

**Review by h-vetinari — COMMENTED** [review]
@jreback For pins, please check OP. Happy to change however you want. Will bump matplotlib pending further input from you (already to 2.2.2?)

File: pandas/tests/plotting/common.py
Comment (jreback) [reply pandas/tests/plotting/common.py]: cc @tacaswell any thoughts here?

**Comment by h-vetinari** [comment]
@jreback 
Checked the MPL releases, and `2.2.2` is barely a year old. Therefore, I bumped only to `2.1.0`.

**Comment by h-vetinari** [comment]
@jreback 
OK, nevermind, I went to matplotlib 2.2.2 after all. By the time pandas 0.24.2 hit, this will be over a year old, and there's just too much variability in the MPL point releases for `2.1.x` so that any starting point in `2.0.1 | 2.0.2 | 2.1.0 | 2.1.1 | 2.1.2` (and to a degree `2.2.0`) would lead to lots of compat code (along the lines of what you didn't like above). And since 2.2.2 was only released 10 days after 2.2.0, I went to 2.2.2 right away.

**Comment by tacaswell** [comment]
Sorry for the Matplotlib API thrashing.  

2.2.x is the LTS version so it is a good choice!

**Comment by h-vetinari** [comment]
@jreback 
I keep running into a weird cython error on the `azure-windows-27` job:
```
running build_ext
pandas._libs.algos: -> [['pandas\\_libs/algos.c']]
Traceback (most recent call last):
  File "setup.py", line 751, in <module>
    **setuptools_kwargs)
  File "C:\Miniconda\lib\site-packages\setuptools\__init__.py", line 143, in setup
    return distutils.core.setup(**attrs)
  File "C:\Miniconda\lib\distutils\core.py", line 148, in setup
    dist.run_commands()
  File "C:\Miniconda\lib\distutils\dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "C:\Miniconda\lib\distutils\dist.py", line 985, in run_command
    cmd_obj.run()
  File "C:\Miniconda\lib\distutils\command\build_ext.py", line 339, in run
    self.build_extensions()
  File "setup.py", line 372, in build_extensions
    self.check_cython_extensions(self.extensions)
  File "setup.py", line 369, in check_cython_extensions
    """.format(src=src))
Exception: Cython-generated file 'pandas\_libs/algos.c' not found.
                Cython is required to compile pandas from a development branch.
                Please install Cython or download a release package of pandas.
```

Bumping cython didn't help (resp. then the 3.5 jobs can't satisfy min). How about we drop the PY2 CI (#24942) first before bumping numpy?

**Comment by WillAyd** [comment]
@h-vetinari can you merge master on this now that Py2 CI jobs are out?

**Comment by h-vetinari** [comment]
> @WillAyd: @h-vetinari can you merge master on this now that Py2 CI jobs are out?

Since #24942 was split up, this PR here still depends on #25752. It would be helpful if you could give the latter one a look. :)

**Comment by h-vetinari** [comment]
Currently rebasing on top of #25752. This is the current set of pins for python/numpy/matplotlib. Comments welcome.

CI job: line number | OS | python | numpy | mpl
-------|-----------------------------|--------|----------|---
`azure-35-compat.yaml:10:` | posix | `python=3.5.*` | `numpy=1.13.3` | -
`azure-36-locale.yaml:10:` | posix | `python=3.6.*` | `numpy=1.14.*` | `matplotlib=2.2.2`
`azure-36-locale_slow.yaml:16:` | posix |  `python=3.6.*` | `numpy=1.15.*`  | `matplotlib=3.0.*` 
`azure-37-locale.yaml:15:` | posix |  `python=3.7.*` | `numpy`  | `matplotlib`
`azure-37-numpydev.yaml:18:` | posix |  `python=3.7.*` | `numpy`  | -
`azure-macos-35.yaml:14:` | macos |  `python=3.5.*` | `numpy=1.13.3` | `matplotlib=2.2.3`
`azure-windows-36.yaml:15:` | win |  `python=3.6.*` | `numpy=1.15.*` | `matplotlib`
`azure-windows-37.yaml:12:` | win |  `python=3.6.*` | `numpy=1.14.*` | `matplotlib=2.2.*`
`travis-36-doc.yaml:23:` | posix |  `python=3.6.*` | `numpy` | `matplotlib`
`travis-36-locale.yaml:15:` | posix | `python=3.6.*` | `numpy` | `matplotlib=3.0.1`
`travis-36-slow.yaml:12:` | posix | `python=3.6.*` | `numpy` | `matplotlib`
`travis-36.yaml:17:` | posix | `python=3.6.*` | `numpy=1.15.*` | `matplotlib=3.0.0`
`travis-37.yaml:10:` | posix | `python=3.7.*` | `numpy` | -

**Comment by jreback** [comment]
see if we can remove 1 or more jobs keeping the same coverage (reasonably) of deps (can be a follow up)
i don’t think it’s a big deal if we only test oldest numpy and then newest
1.14/1.15 or pretty much compat with 1.16

**Comment by h-vetinari** [comment]
Well, I didn't design the CI jobs - there's OSes, numpies, locales, DB tests. Not sure we should reduce much more, but from the table above, I'd get rid of `travis-3.6-locale`.

Also, we still need to actually merge #25752.

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: this is not needed

File: environment.yml
Comment (jreback) [inline environment.yml]: this seems too high of a version here for numexpr compared to the above

File: pandas/tests/plotting/common.py
Comment (jreback) [inline pandas/tests/plotting/common.py]: shouldn't you just make this .mpl_ge_2_2_2

File: pandas/tests/plotting/test_datetimelike.py
Comment (jreback) [inline pandas/tests/plotting/test_datetimelike.py]: see my above comment, you should just be able to do : ``not self.mpl_ge_3_0_3`` to get this condition

File: requirements-dev.txt
Comment (jreback) [inline requirements-dev.txt]: same change here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
https://travis-ci.org/pandas-dev/pandas/jobs/510922188 appears to be failing

File: environment.yml
Comment (h-vetinari) [reply environment.yml]: This was added in #24015 and does not affect the lowest supported version

File: pandas/tests/plotting/common.py
Comment (h-vetinari) [reply pandas/tests/plotting/common.py]: It was intentional because the changes in `test/plotting/test_datetimelike` affected *only* 2.2.2, but after the bump to 2.2.2, this can now be formulated also with `_ge_`.

File: doc/source/whatsnew/v0.25.0.rst
Comment (h-vetinari) [inline doc/source/whatsnew/v0.25.0.rst]: Not sure if that's what you had in mind in your [review](https://github.com/pandas-dev/pandas/pull/25752#issuecomment-476008611), but I think this is actually a good overview that:
* represents the way this is handled in practice
* is hard to find otherwise

Also, going through this exercise made me realise that by requiring the removal of the `travis-27` job, `pandas-gbq` was completely untested. I've ended up using `travis-36-locale` as a sort of compat-job for all the libraries that are not already covered by `azure-35-compat`.

Some other versions had no compat testing whatsoever (which does not fit with specifying minimum versions for them in `install.rst`), so I've added some.

**Review by h-vetinari — COMMENTED** [review]
Looking a bit more after which minimum versions were enforced, I found lots more cruft to remove.

Also, in trying to find a better formulation for the whatsnew entry, I found that we had still left some holes in the CI by removing the `travis-27` job completely.

PTAL

File: ci/deps/travis-36-cov.yaml
Comment (h-vetinari) [inline ci/deps/travis-36-cov.yaml]: I added a pinned version in `travis-36-locale` and an unpinned one here (as this had fallen off the radar completely)

File: ci/deps/travis-36-locale.yaml
Comment (h-vetinari) [inline ci/deps/travis-36-locale.yaml]: This was the reason for https://github.com/pandas-dev/pandas/pull/25554#issuecomment-476183483.

I also shifted it out of the `slow` build where failures are not immediately apparent.

File: .travis.yml
Comment (h-vetinari) [inline .travis.yml]: I made the name of the coverage build more explicit  `travis-36` -> `travis-36-cov`. It had some pins it should not have had, and it's easy to forget that when not being reminded by the name (same for the `travis-36-doc` build).

**Review by h-vetinari — COMMENTED** [review]
Some more comments about the CI changes.

File: ci/deps/azure-35-compat.yaml
Comment (jreback) [inline ci/deps/azure-35-compat.yaml]: I don't think this actually matters  but ok

File: ci/deps/travis-36-locale.yaml
Comment (jreback) [inline ci/deps/travis-36-locale.yaml]: i am suprised this doesn't break anything, we should be using a higher version for testting (we may be elsewhere),  maybe add an issue for consideration later

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: this is too verbose. pls separate into 2 tables, the basic deps (numpy, pytz, bottleneck, numexpr); make the table like we had it exactly

you can make a table for other things, but i am going to limit that list.

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: remove beautifulsoup, blosc, gcfs, junja2, lxml, psycopg2, pymsql, s3fs, cython and pytest

**Review by jreback — CHANGES_REQUESTED** [review]
ok looks good. don't make *any* more changes, except for chopping the whatsnew as indicated. ping on green.

File: ci/deps/azure-35-compat.yaml
Comment (h-vetinari) [reply ci/deps/azure-35-compat.yaml]: It does not, but I made it consistent across jobs.

File: ci/deps/travis-36-locale.yaml
Comment (h-vetinari) [reply ci/deps/travis-36-locale.yaml]: All the pinned libraries here have at least one unpinned counterpart in another job.

File: doc/source/whatsnew/v0.25.0.rst
Comment (h-vetinari) [reply doc/source/whatsnew/v0.25.0.rst]: pytest was explicitly added to this table by #23519.

I added all the libraries from https://pandas-docs.github.io/pandas-docs-travis/install.html#dependencies. I think this is a worthwhile overview to have *somewhere* (maybe not the whatsnew). Thoughts, @datapythonista?

**Review by h-vetinari — COMMENTED** [review]
Will follow up later today, no time right now.

File: doc/source/whatsnew/v0.25.0.rst
Comment (h-vetinari) [reply doc/source/whatsnew/v0.25.0.rst]: I kept pytest due to #23519.

I kept the structure of the table because the distinction between the kind of requirement (hard requirement vs. requirement if installed) is clearer like this, and fits with the other table.

**Review by h-vetinari — COMMENTED** [review]
PTAL

**Comment by h-vetinari** [comment]
@jreback 
Argh, now we're right back to the errors I had fixed with the regex stuff that you reverted against my request. #25874

File: pandas/tests/test_sorting.py
Comment (jreback) [inline pandas/tests/test_sorting.py]: this fails the wheel building CI. pls don't revert this, I don't want to debug this problem. Just make this MUCH simpler so it will pretty much always pass.

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: this is a misleading comment for pytest. its required if you run tests

Require if installed is very misleading, remove it

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: remove the comment column.

**Review by jreback — CHANGES_REQUESTED** [review]
pls just make the changes I indicated. otherwise this iteration cycle will continue to extend.

File: pandas/tests/test_sorting.py
Comment (h-vetinari) [reply pandas/tests/test_sorting.py]: It will *not* fail the wheel building CI. I actually checked the wheel builder failures from #25867 before giving my [review](https://github.com/pandas-dev/pandas/pull/25874#pullrequestreview-218710820) in #25874.

In case you don't have time to read the review, the original code in #25752 has an `if _np_version_under1p14 else`-switch that turned out to be wrong, while the regex here only used a `|`-conjunction.

> [...] I don't want to debug this problem.

Neither did I, but I've had to for a passing CI. Please have a look at the actual regexes. What I've written is a proper superset of the current regexes (again, now that the switch has been turned into `"|"`).

**Review by h-vetinari — COMMENTED** [review]
Updated. Please take the time to read my comment about the regex.

File: pandas/tests/test_sorting.py
Comment (h-vetinari) [reply pandas/tests/test_sorting.py]: > Just make this MUCH simpler so it will pretty much always pass.

Done.

**Comment by jreback** [comment]
> It will not fail the wheel building CI. I actually checked the wheel builder failures from #25867 before giving my review in #25874.

not sure what you are talking about. it DID fail. The reverted does not. If you fixed it great. I want to avoid having broken CI's endlessly.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Comment by jreback** [comment]
@h-vetinari looks like this broke the gbq tests; these are pretty funky and I don't think are tested on  actual branches (thoguh they used to be AFAIR): see https://travis-ci.org/pandas-dev/pandas/jobs/512494733

so pls do a PR to disable for now.

**Comment by h-vetinari** [comment]
> @jreback: not sure what you are talking about. it DID fail. The reverted does not. If you fixed it great. I want to avoid having broken CI's endlessly.

I did fail, and I had fixed it (not by reverting exactly, but then I've tried to explain that already twice). Now it's more simplified anyway, which is IMO anyway a better approach for matching these error messages.

**Pull request merged**
